storage/shard.go

```diff
@@ -1040,28 +1153,44 @@ func (s *dbShard) insertSeriesBatch(inserts []dbShardInsert) error {
 		return nil
 	}
 
-	// Perform any pending writes or pending retrieved blocks outside of lock
+	// Perform any indexing, pending writes or pending retrieved blocks outside of lock
 	ctx := s.contextPool.Get()
 	for i := range inserts {
 		var (
-			entry       = inserts[i].entry
-			readOrWrite = false
+			entry           = inserts[i].entry
+			releaseEntryRef = false
 		)
-		switch {
-		case inserts[i].opts.hasPendingWrite:
-			readOrWrite = true
+
+		if inserts[i].opts.hasPendingWrite {
+			releaseEntryRef = true
 			write := inserts[i].opts.pendingWrite
 			err := entry.series.Write(ctx, write.timestamp, write.value,
 				write.unit, write.annotation)
 			if err != nil {
 				s.metrics.insertAsyncWriteErrors.Inc(1)
 			}
-		case inserts[i].opts.hasPendingRetrievedBlock:
-			readOrWrite = true
+		}
+
+		if inserts[i].opts.hasPendingIndexing {
+			pendingIndex := inserts[i].opts.pendingIndex
+			releaseEntryRef = true
+			// only index any entry that hasn't crossed the nextIndexTime
+			if entry.needsIndexUpdate(pendingIndex.timestamp) {
+				// NB(prateek): we index entries asynchronously, so if we are going to
+				// index the entry, we inc an extra ref. It is released by the indexing
+				// function callback - `entry.OnIndexFinalize`.
+				entry.incrementReaderWriterCount()
+				pendingIndex.fn(entry.series.ID(), entry.series.Tags(), entry)
```

Comment:
Is there any reason you need to pass the fn into the shard insert queue?  Couldn't you instead just `s.reverseIndex.Write(id, tags, entry)` here?

Reply:
Yep, you're right. Will do

Reply:
done